File: server/src/main/java/net/lumamc/web/model/BlockbenchModel.java
package net.lumamc.web.model;

import com.google.gson.JsonArray;
import com.google.gson.JsonElement;
import com.google.gson.JsonObject;
import com.google.gson.JsonParser;

import java.io.IOException;
import java.nio.charset.StandardCharsets;
import java.nio.file.Files;
import java.nio.file.Path;
import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;


public final class BlockbenchModel {

    public static final class Triangle {
        public final double[] a, b, c;
        public final double[] uvA, uvB, uvC;

        Triangle(double[] a, double[] b, double[] c,
                 double[] uvA, double[] uvB, double[] uvC) {
            this.a = a;
            this.b = b;
            this.c = c;
            this.uvA = uvA;
            this.uvB = uvB;
            this.uvC = uvC;
        }
    }

    private final List<Triangle> triangles = new ArrayList<>();
    private final double uvWidth, uvHeight;

    private BlockbenchModel(double uvWidth, double uvHeight) {
        this.uvWidth = uvWidth;
        this.uvHeight = uvHeight;
    }

    public List<Triangle> triangles() {
        return triangles;
    }

    public double uvWidth() {
        return uvWidth;
    }

    public double uvHeight() {
        return uvHeight;
    }

    public static BlockbenchModel load(Path file) throws IOException {
        return parse(new String(Files.readAllBytes(file), StandardCharsets.UTF_8));
    }

    public static BlockbenchModel parse(String jsonText) {
        JsonObject root = JsonParser.parseString(jsonText).getAsJsonObject();

        double uvW = 64, uvH = 64;
        if (root.has("resolution")) {
            JsonObject res = root.getAsJsonObject("resolution");
            uvW = optDouble(res, "width", 64);
            uvH = optDouble(res, "height", 64);
        }
        BlockbenchModel model = new BlockbenchModel(uvW, uvH);

        Map<String, JsonObject> elements = new HashMap<>();
        if (root.has("elements")) {
            for (JsonElement e : root.getAsJsonArray("elements")) {
                JsonObject eo = e.getAsJsonObject();
                elements.put(eo.get("uuid").getAsString(), eo);
            }
        }

        Map<String, JsonObject> groups = new HashMap<>();
        if (root.has("groups")) {
            for (JsonElement g : root.getAsJsonArray("groups")) {
                JsonObject go = g.getAsJsonObject();
                groups.put(go.get("uuid").getAsString(), go);
            }
        }

        // Find each element's chain of parent groups, innermost first.
        Map<String, List<JsonObject>> ancestry = new HashMap<>();
        if (root.has("outliner")) {
            for (JsonElement node : root.getAsJsonArray("outliner")) {
                walkOutliner(node, new ArrayList<>(), groups, elements, ancestry);
            }
        }

        for (Map.Entry<String, JsonObject> entry : elements.entrySet()) {
            JsonObject e = entry.getValue();
            String type = e.has("type") ? e.get("type").getAsString() : "cube";
            if (!"cube".equals(type)) {
                continue;
            }
            List<JsonObject> chain = ancestry.getOrDefault(entry.getKey(), new ArrayList<>());
            model.addCube(e, chain);
        }

        return model;
    }

    private static void walkOutliner(JsonElement node, List<JsonObject> ancestorsInnermostFirst, Map<String, JsonObject> groups, Map<String, JsonObject> elements, Map<String, List<JsonObject>> out) {
        if (node.isJsonPrimitive()) {
            String uuid = node.getAsString();
            if (elements.containsKey(uuid)) {
                out.put(uuid, new ArrayList<>(ancestorsInnermostFirst));
            }
            return;
        }

        JsonObject n = node.getAsJsonObject();
        JsonObject group = groups.get(n.get("uuid").getAsString());

        List<JsonObject> nextAncestors = new ArrayList<>();
        if (group != null) {
            nextAncestors.add(group);
        }
        nextAncestors.addAll(ancestorsInnermostFirst);

        if (n.has("children")) {
            for (JsonElement child : n.getAsJsonArray("children")) {
                walkOutliner(child, nextAncestors, groups, elements, out);
            }
        }
    }

    private static final String[] FACE_NAMES =
            {"north", "south", "west", "east", "up", "down"};

    private void addCube(JsonObject e, List<JsonObject> groupChain) {
        double inflate = optDouble(e, "inflate", 0.0);
        double[] from = vec3(e.get("from"));
        double[] to = vec3(e.get("to"));

        double x0 = from[0] - inflate, y0 = from[1] - inflate, z0 = from[2] - inflate;
        double x1 = to[0] + inflate, y1 = to[1] + inflate, z1 = to[2] + inflate;

        JsonObject faces = e.getAsJsonObject("faces");

        for (String face : FACE_NAMES) {
            if (!faces.has(face)) {
                continue;
            }

            JsonObject fd = faces.getAsJsonObject(face);
            // Faces with no texture are not drawn, like the hitbox cube.
            if (!fd.has("texture") || fd.get("texture").isJsonNull()) {
                continue;
            }
            if (!fd.has("uv")) {
                continue;
            }

            JsonArray uv = fd.getAsJsonArray("uv");
            double u1 = uv.get(0).getAsDouble(), v1 = uv.get(1).getAsDouble();
            double u2 = uv.get(2).getAsDouble(), v2 = uv.get(3).getAsDouble();

            double[][] corners = faceCorners(face, x0, y0, z0, x1, y1, z1);

            // Rotate by the element first, then up through the group chain.
            double[] elemOrigin = vec3OrZero(e, "origin");
            double[] elemRotation = vec3OrZero(e, "rotation");
            for (int k = 0; k < 4; k++) {
                corners[k] = rotateAround(corners[k], elemOrigin, elemRotation);
                for (JsonObject g : groupChain) {
                    corners[k] = rotateAround(corners[k], vec3(g.get("origin")), vec3OrZero(g, "rotation"));
                }
            }

            double[] tl = {u1, v1}, tr = {u2, v1}, br = {u2, v2}, bl = {u1, v2};
            triangles.add(new Triangle(corners[0], corners[1], corners[2], tl, tr, br));
            triangles.add(new Triangle(corners[0], corners[2], corners[3], tl, br, bl));
        }
    }

    // Face corners in the order top left, top right, bottom right, bottom left
    // so they line up with the uv corners.
    private static double[][] faceCorners(String face, double x0, double y0, double z0, double x1, double y1, double z1) {
        return switch (face) {
            case "north" -> new double[][]{{x1, y1, z0}, {x0, y1, z0}, {x0, y0, z0}, {x1, y0, z0}};
            case "south" -> new double[][]{{x0, y1, z1}, {x1, y1, z1}, {x1, y0, z1}, {x0, y0, z1}};
            case "west" -> new double[][]{{x0, y1, z0}, {x0, y1, z1}, {x0, y0, z1}, {x0, y0, z0}};
            case "east" -> new double[][]{{x1, y1, z1}, {x1, y1, z0}, {x1, y0, z0}, {x1, y0, z1}};
            case "up" -> new double[][]{{x0, y1, z0}, {x1, y1, z0}, {x1, y1, z1}, {x0, y1, z1}};
            case "down" -> new double[][]{{x0, y0, z1}, {x1, y0, z1}, {x1, y0, z0}, {x0, y0, z0}};
            default -> throw new IllegalArgumentException("face " + face);
        };
    }

    // Rotate a point around an origin using xyz euler angles in degrees.
    private static double[] rotateAround(double[] p, double[] origin, double[] rotDeg) {
        if (rotDeg[0] == 0 && rotDeg[1] == 0 && rotDeg[2] == 0) {
            return p;
        }

        double x = p[0] - origin[0], y = p[1] - origin[1], z = p[2] - origin[2];
        double rx = Math.toRadians(rotDeg[0]);
        double ry = Math.toRadians(rotDeg[1]);
        double rz = Math.toRadians(rotDeg[2]);
        double cx = Math.cos(rx), sx = Math.sin(rx);
        double cy = Math.cos(ry), sy = Math.sin(ry);
        double cz = Math.cos(rz), sz = Math.sin(rz);

        double ny = y * cx - z * sx, nz = y * sx + z * cx;
        y = ny;
        z = nz;

        double nx = x * cy + z * sy;
        nz = -x * sy + z * cy;
        x = nx;
        z = nz;

        nx = x * cz - y * sz;
        ny = x * sz + y * cz;
        x = nx;
        y = ny;

        return new double[]{x + origin[0], y + origin[1], z + origin[2]};
    }

    private static double optDouble(JsonObject o, String key, double def) {
        if (o.has(key) && !o.get(key).isJsonNull()) {
            return o.get(key).getAsDouble();
        }
        return def;
    }

    private static double[] vec3(JsonElement el) {
        JsonArray a = el.getAsJsonArray();
        return new double[]{a.get(0).getAsDouble(), a.get(1).getAsDouble(), a.get(2).getAsDouble()};
    }

    private static double[] vec3OrZero(JsonObject o, String key) {
        if (!o.has(key) || o.get(key).isJsonNull()) {
            return new double[]{0, 0, 0};
        }
        return vec3(o.get(key));
    }
}

File: server/src/main/kotlin/net/lumamc/web/model/BlockbenchModel.kt
package net.lumamc.web.model

import com.google.gson.JsonElement
import com.google.gson.JsonObject
import com.google.gson.JsonParser
import java.awt.image.BufferedImage
import java.io.ByteArrayInputStream
import java.nio.file.Files
import java.nio.file.Path
import java.util.Base64
import javax.imageio.ImageIO
import kotlin.math.cos
import kotlin.math.sin

class BlockbenchModel private constructor(
    private val uvWidth: Double,
    private val uvHeight: Double,
    private val textures: List<Texture>,
    private val skinTexture: Int
) {

    class Texture(
        @JvmField val name: String,
        @JvmField val id: String?,
        @JvmField val uuid: String?,
        @JvmField val uvWidth: Double,
        @JvmField val uvHeight: Double,
        @JvmField val image: BufferedImage?
    )

    class Triangle(
        @JvmField val a: DoubleArray,
        @JvmField val b: DoubleArray,
        @JvmField val c: DoubleArray,
        @JvmField val uvA: DoubleArray,
        @JvmField val uvB: DoubleArray,
        @JvmField val uvC: DoubleArray,
        // Index into textures(), or -1 when the face did not name a texture we know about.
        @JvmField val texture: Int
    )

    private val triangles = mutableListOf<Triangle>()

    fun triangles(): List<Triangle> = triangles

    fun textures(): List<Texture> = textures

    // Index of the texture the player skin is painted onto, or -1 when the model has no such slot.
    fun skinTexture(): Int = skinTexture

    fun uvWidth(): Double = uvWidth

    fun uvHeight(): Double = uvHeight

    private fun addCube(element: JsonObject, groupChain: List<JsonObject>) {
        val inflate = optDouble(element, "inflate", 0.0)
        val from = vec3(element.get("from"))
        val to = vec3(element.get("to"))

        val x0 = from[0] - inflate
        val y0 = from[1] - inflate
        val z0 = from[2] - inflate
        val x1 = to[0] + inflate
        val y1 = to[1] + inflate
        val z1 = to[2] + inflate

        val faces = element.getAsJsonObject("faces") ?: return
        val elementOrigin = vec3OrZero(element, "origin")
        val elementRotation = vec3OrZero(element, "rotation")
        for (face in FACE_NAMES) {
            if (!faces.has(face)) {
                continue
            }

            val faceData = faces.getAsJsonObject(face)
            if (!faceData.has("uv")) {
                continue
            }
            val texture = textureIndex(faceData)
            if (texture == NO_TEXTURE && textures.isNotEmpty()) {
                // The face is deliberately untextured.
                continue
            }

            val uv = faceData.getAsJsonArray("uv")
            val u1 = uv[0].asDouble
            val v1 = uv[1].asDouble
            val u2 = uv[2].asDouble
            val v2 = uv[3].asDouble
            val corners = faceCorners(face, x0, y0, z0, x1, y1, z1)

            for (index in corners.indices) {
                corners[index] = rotateAround(corners[index], elementOrigin, elementRotation)
                for (group in groupChain) {
                    corners[index] = rotateAround(
                        corners[index],
                        vec3OrZero(group, "origin"),
                        vec3OrZero(group, "rotation")
                    )
                }
            }

            val topLeft = doubleArrayOf(u1, v1)
            val topRight = doubleArrayOf(u2, v1)
            val bottomRight = doubleArrayOf(u2, v2)
            val bottomLeft = doubleArrayOf(u1, v2)
            triangles += Triangle(corners[0], corners[1], corners[2], topLeft, topRight, bottomRight, texture)
            triangles += Triangle(corners[0], corners[2], corners[3], topLeft, bottomRight, bottomLeft, texture)
        }
    }

    private fun textureIndex(faceData: JsonObject): Int {
        val reference = faceData.get("texture")
        if (reference == null || reference.isJsonNull || !reference.isJsonPrimitive) {
            return NO_TEXTURE
        }

        val primitive = reference.asJsonPrimitive
        if (primitive.isNumber) {
            val index = primitive.asInt
            return if (index in textures.indices) index else NO_TEXTURE
        }

        val key = primitive.asString
        val index = textures.indexOfFirst { it.uuid == key || it.id == key }
        return if (index >= 0) index else NO_TEXTURE
    }

    companion object {
        private const val NO_TEXTURE = -1
        private val FACE_NAMES = arrayOf("north", "south", "west", "east", "up", "down")
        private val SKIN_NAME_HINTS = arrayOf("skin", "player", "_you", "steve")

        @JvmStatic
        fun load(file: Path): BlockbenchModel = parse(Files.readString(file))

        @JvmStatic
        fun parse(jsonText: String): BlockbenchModel {
            val root = JsonParser.parseString(jsonText).asJsonObject

            var uvWidth = 64.0
            var uvHeight = 64.0
            if (root.has("resolution")) {
                val resolution = root.getAsJsonObject("resolution")
                uvWidth = optDouble(resolution, "width", 64.0)
                uvHeight = optDouble(resolution, "height", 64.0)
            }

            val textures = parseTextures(root, uvWidth, uvHeight)
            val model = BlockbenchModel(uvWidth, uvHeight, textures, findSkinTexture(textures))

            val elements = mutableMapOf<String, JsonObject>()
            if (root.has("elements")) {
                for (element in root.getAsJsonArray("elements")) {
                    val elementObject = element.asJsonObject
                    if (!isVisible(elementObject)) {
                        continue
                    }
                    elements[elementObject.get("uuid").asString] = elementObject
                }
            }

            val groups = mutableMapOf<String, JsonObject>()
            if (root.has("groups")) {
                for (group in root.getAsJsonArray("groups")) {
                    val groupObject = group.asJsonObject
                    groups[groupObject.get("uuid").asString] = groupObject
                }
            }

            val ancestry = mutableMapOf<String, List<JsonObject>>()
            val hidden = mutableSetOf<String>()
            if (root.has("outliner")) {
                for (node in root.getAsJsonArray("outliner")) {
                    walkOutliner(node, emptyList(), groups, elements, ancestry, hidden)
                }
            }

            for ((id, element) in elements) {
                if (id in hidden) {
                    continue
                }
                val type = if (element.has("type")) element.get("type").asString else "cube"
                if (type == "cube") {
                    model.addCube(element, ancestry[id].orEmpty())
                }
            }

            return model
        }

        private fun parseTextures(root: JsonObject, uvWidth: Double, uvHeight: Double): List<Texture> {
            if (!root.has("textures")) {
                return emptyList()
            }

            return root.getAsJsonArray("textures").map { entry ->
                val texture = entry.asJsonObject
                Texture(
                    optString(texture, "name").orEmpty(),
                    optString(texture, "id"),
                    optString(texture, "uuid"),
                    optDouble(texture, "uv_width", optDouble(texture, "width", uvWidth)),
                    optDouble(texture, "uv_height", optDouble(texture, "height", uvHeight)),
                    decodeTexture(optString(texture, "source"))
                )
            }
        }

        private fun decodeTexture(source: String?): BufferedImage? {
            if (source.isNullOrBlank()) {
                return null
            }

            val payload = if (source.startsWith("data:")) {
                val comma = source.indexOf(',')
                if (comma < 0) return null else source.substring(comma + 1)
            } else {
                // A plain path to a file next to the model, which we do not follow.
                return null
            }

            return try {
                ImageIO.read(ByteArrayInputStream(Base64.getMimeDecoder().decode(payload)))
            } catch (e: Exception) {
                null
            }
        }

        private fun findSkinTexture(textures: List<Texture>): Int {
            if (textures.isEmpty()) {
                return NO_TEXTURE
            }
            if (textures.size == 1) {
                return 0
            }

            val skinSized = textures.indices.filter {
                textures[it].uvWidth == 64.0 && (textures[it].uvHeight == 64.0 || textures[it].uvHeight == 32.0)
            }
            val named = textures.indices.filter { index ->
                val name = textures[index].name.lowercase()
                SKIN_NAME_HINTS.any { name.contains(it) }
            }

            return skinSized.firstOrNull { it in named }
                ?: named.firstOrNull()
                ?: skinSized.singleOrNull()
                // Nothing looks like a skin, but a texture without an image can only be one.
                ?: textures.indices.firstOrNull { textures[it].image == null }
                ?: NO_TEXTURE
        }

        private fun walkOutliner(
            node: JsonElement,
            ancestorsInnermostFirst: List<JsonObject>,
            groups: Map<String, JsonObject>,
            elements: Map<String, JsonObject>,
            output: MutableMap<String, List<JsonObject>>,
            hidden: MutableSet<String>
        ) {
            if (node.isJsonPrimitive) {
                val id = node.asString
                if (elements.containsKey(id)) {
                    output[id] = ancestorsInnermostFirst.toList()
                }
                return
            }

            val nodeObject = node.asJsonObject
            // models either keep their groups in a separate list or inline them in the outliner.
            val group = groups[nodeObject.get("uuid").asString] ?: nodeObject
            if (!isVisible(group)) {
                collectIds(nodeObject, hidden)
                return
            }

            val nextAncestors = buildList {
                add(group)
                addAll(ancestorsInnermostFirst)
            }

            if (nodeObject.has("children")) {
                for (child in nodeObject.getAsJsonArray("children")) {
                    walkOutliner(child, nextAncestors, groups, elements, output, hidden)
                }
            }
        }

        private fun collectIds(node: JsonElement, output: MutableSet<String>) {
            if (node.isJsonPrimitive) {
                output += node.asString
                return
            }

            val nodeObject = node.asJsonObject
            if (nodeObject.has("children")) {
                for (child in nodeObject.getAsJsonArray("children")) {
                    collectIds(child, output)
                }
            }
        }

        private fun isVisible(node: JsonObject): Boolean {
            val exported = !node.has("export") || node.get("export").isJsonNull || node.get("export").asBoolean
            val visible = !node.has("visibility") ||
                node.get("visibility").isJsonNull ||
                node.get("visibility").asBoolean
            return exported && visible
        }

        private fun faceCorners(
            face: String,
            x0: Double,
            y0: Double,
            z0: Double,
            x1: Double,
            y1: Double,
            z1: Double
        ): Array<DoubleArray> = when (face) {
            "north" -> arrayOf(
                doubleArrayOf(x1, y1, z0), doubleArrayOf(x0, y1, z0),
                doubleArrayOf(x0, y0, z0), doubleArrayOf(x1, y0, z0)
            )
            "south" -> arrayOf(
                doubleArrayOf(x0, y1, z1), doubleArrayOf(x1, y1, z1),
                doubleArrayOf(x1, y0, z1), doubleArrayOf(x0, y0, z1)
            )
            "west" -> arrayOf(
                doubleArrayOf(x0, y1, z0), doubleArrayOf(x0, y1, z1),
                doubleArrayOf(x0, y0, z1), doubleArrayOf(x0, y0, z0)
            )
            "east" -> arrayOf(
                doubleArrayOf(x1, y1, z1), doubleArrayOf(x1, y1, z0),
                doubleArrayOf(x1, y0, z0), doubleArrayOf(x1, y0, z1)
            )
            "up" -> arrayOf(
                doubleArrayOf(x0, y1, z0), doubleArrayOf(x1, y1, z0),
                doubleArrayOf(x1, y1, z1), doubleArrayOf(x0, y1, z1)
            )
            "down" -> arrayOf(
                doubleArrayOf(x0, y0, z1), doubleArrayOf(x1, y0, z1),
                doubleArrayOf(x1, y0, z0), doubleArrayOf(x0, y0, z0)
            )
            else -> throw IllegalArgumentException("face $face")
        }

        private fun rotateAround(point: DoubleArray, origin: DoubleArray, rotationDegrees: DoubleArray): DoubleArray {
            if (rotationDegrees.all { it == 0.0 }) {
                return point
            }

            var x = point[0] - origin[0]
            var y = point[1] - origin[1]
            var z = point[2] - origin[2]
            val rotationX = Math.toRadians(rotationDegrees[0])
            val rotationY = Math.toRadians(rotationDegrees[1])
            val rotationZ = Math.toRadians(rotationDegrees[2])
            val cosineX = cos(rotationX)
            val sineX = sin(rotationX)
            val cosineY = cos(rotationY)
            val sineY = sin(rotationY)
            val cosineZ = cos(rotationZ)
            val sineZ = sin(rotationZ)

            var nextY = y * cosineX - z * sineX
            var nextZ = y * sineX + z * cosineX
            y = nextY
            z = nextZ

            var nextX = x * cosineY + z * sineY
            nextZ = -x * sineY + z * cosineY
            x = nextX
            z = nextZ

            nextX = x * cosineZ - y * sineZ
            nextY = x * sineZ + y * cosineZ

            return doubleArrayOf(nextX + origin[0], nextY + origin[1], z + origin[2])
        }

        private fun optDouble(jsonObject: JsonObject, key: String, default: Double): Double {
            return if (jsonObject.has(key) && !jsonObject.get(key).isJsonNull) {
                jsonObject.get(key).asDouble
            } else {
                default
            }
        }

        private fun optString(jsonObject: JsonObject, key: String): String? {
            return if (jsonObject.has(key) && !jsonObject.get(key).isJsonNull) {
                jsonObject.get(key).asString
            } else {
                null
            }
        }

        private fun vec3(element: JsonElement): DoubleArray {
            val array = element.asJsonArray
            return doubleArrayOf(array[0].asDouble, array[1].asDouble, array[2].asDouble)
        }

        private fun vec3OrZero(jsonObject: JsonObject, key: String): DoubleArray {
            return if (!jsonObject.has(key) || jsonObject.get(key).isJsonNull) {
                doubleArrayOf(0.0, 0.0, 0.0)
            } else {
                vec3(jsonObject.get(key))
            }
        }
    }
}
